pgjdbc/src/main/java/org/postgresql/jdbcurlresolver/JdbcUrlParser.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2022, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.jdbcurlresolver;
+
+import org.postgresql.PGProperty;
+import org.postgresql.util.URLCoder;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+class JdbcUrlParser {
+
+  // position options
+  private enum SECTION {
+    USER, HOST, DATABASE, ARGUMENT
+  }
+
+  // input
+  private final String url;
+
+  // work state
+  private @Nullable SECTION currentSection;
+  private String urlTemp = "";
+  private @Nullable String urlHostPort;
+
+  // results
+  private @Nullable String urlUser;
+  private @Nullable String urlPass;
+  private @Nullable String urlHost;
+  private @Nullable String urlPort;
+  private @Nullable String urlDatabase;
+  private @Nullable String urlArguments;
+
+  //
+  JdbcUrlParser(String url) {
+    this.url = url;
+  }
+
+  void parse(Properties arguments, Properties values) throws JdbcUrlResolverFatalException {
+    // split url into sections: user, password, hosts, database, arguments
+    parse10UrlPrefix();
+    parse20User();
+    parse30Password();
+    parse40Host();
+    parse50Database();
+    parse60Argument();
+    // split host section (in case of multiple hosts)
+    parse70HostPort();
+    // result - priority 1: URL arguments
+    result10Argument(arguments);
+    // result - priority 2: URL values
+    result20Url(values);
+  }
+
+  private void saveSection(String value) {
+    if (currentSection == null) {
+      throw new RuntimeException(String.format("bug in code, unhandled currentSection [%s]", currentSection));
+    }
+    if (value.isEmpty()) {
+      return;
+    }
+    switch (currentSection) {
+      case USER:
+        urlUser = value;
+        break;
+      case HOST:
+        urlHostPort = value;
+        break;
+      case DATABASE:
+        urlDatabase = value;
+        break;
+      case ARGUMENT:
+        urlArguments = value;
+        break;
+      default:
+        throw new RuntimeException(String.format("bug in code, unhandled currentSection [%s]", currentSection.name()));
+    }
+  }
+
+  private void parse10UrlPrefix() throws JdbcUrlResolverFatalException {
```

Comment:
What is the reasoning behind having numbers in the names of these functions?

Reply:
For me, it gives the code clarity and readability. Can be renamed if needed